File: RemoveNNodeFromEnd.java
// Time Complexity : O(n)
// Space Complexity : O(1)
public class RemoveNNodeFromEnd {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null)
            return head;
        
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        int len = 0;
        ListNode fast = dummy, slow = dummy;
        // move fast pointer at the n
        while(n >=0)
        {
            fast = fast.next;
            n-- ;
        }
        // move fast pointer until it becomes null
        while(fast != null)
        {
            fast = fast.next;
            slow = slow.next;
        }
        // slow pointer is one before the node to be removed
        slow.next = slow.next.next;
        return dummy.next;
   }
}

File: Remove_N_NodeFromEnd.py
# Time Complexity O(n) n is the number of elements in linked list
# Space complexity O(1)

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        aheadPtr = head
        slowPtr = head
        count = 0
        while aheadPtr:
            if count <= n:
                aheadPtr = aheadPtr.next
                if aheadPtr == None and count < n:
                    return head.next
                count += 1

            else:
                aheadPtr = aheadPtr.next
                slowPtr = slowPtr.next

        slowPtr.next = slowPtr.next.next
        return head
